lib/util-badsalt-chars.c

```diff
@@ -0,0 +1,35 @@
+/* Copyright (C) 2021 Björn Esser <besser82@fedoraproject.org>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/* Table of characters that are invalid for MFC salt-strings.  */
+
+#include "crypt-port.h"
+
+const unsigned char badsalt_chars[166] =
+  "\x01\x02\x03\x04\x05\x06\x07\x08\x0a\x0b\x0c\x0d\x0e\x0f"
+  "\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f"
+  " !*:;\\\x7f"
```

Comment:
So you block a few more printable characters here. The previous list of just two characters to block had solid rationale: those characters wouldn't work right in an `/etc/shadow` entry _anyway_. Now you're blocking characters that would actually work inside a salt. I can see how those are potentially somewhat problematic, but I am not convinced this is obviously the right thing to do. I don't object, though.

For the control characters and the 8-bit range, they're so numerous that maybe checking for them with `strcspn` would have measurable performance impact. Not high and usually negligible, but measurable. With so much effort put into optimizing some of the password hashing implementations, it can be a bit silly to use inefficient constructs in implementing the API. So, while I like how this patch doesn't change the code too much (only replaces the string constant and moves the check inside a function and uses that function twice), maybe it'd be more appropriate to be blocking below 0x20 and above 0x7e with `if` inside a loop. Or choose to not block those.

Reply:
Addressed in rebased commit.